File: alice/paskills/common/vcs/src/main/java/VcsUtils.java
package ru.yandex.alice.paskills.common.vcs;

import java.io.IOException;
import java.net.URL;
import java.util.Enumeration;
import java.util.Objects;
import java.util.jar.Attributes;
import java.util.jar.Manifest;

public class VcsUtils {

    private final String version;
    public static final VcsUtils INSTANCE = new VcsUtils();

    private VcsUtils() {
        String ver = null;
        try {
            Enumeration<URL> resources = getClass().getClassLoader()
                    .getResources("META-INF/MANIFEST.MF");
            while (resources.hasMoreElements()) {
                Manifest manifest = new Manifest(resources.nextElement().openStream());

                Attributes attributes = manifest.getMainAttributes();
                String value = attributes.getValue("Arcadia-Source-Last-Change");
                if (value != null && !value.isEmpty()) {
                    ver = value;
                    break;
                }

            }
        } catch (IOException ignored) {

        }

        this.version = Objects.requireNonNullElse(ver, "");
    }

    public String getVersion() {
        return version;
    }
}

File: alice/kronstadt/core/src/main/java/VcsUtils.kt
package ru.yandex.alice.kronstadt.core

import org.springframework.stereotype.Component
import java.io.IOException
import java.util.jar.Manifest

@Component
open class VcsUtils internal constructor() : VersionProvider {
    override val version: String
    override val branch: String
    override val tag: String

    init {
        var ver: String? = null
        var branchValue: String? = null
        var tagValue: String? = null
        try {
            val resources = javaClass.classLoader.getResources("META-INF/MANIFEST.MF")
            while (resources.hasMoreElements()) {
                val manifest = Manifest(resources.nextElement().openStream())
                val attributes = manifest.mainAttributes
                if (!attributes.getValue("Arcadia-Source-Last-Change").isNullOrEmpty()) {
                    ver = attributes.getValue("Arcadia-Source-Last-Change")
                }
                if (!attributes.getValue("Branch").isNullOrEmpty()) {
                    branchValue = attributes.getValue("Branch")
                }
                if (!attributes.getValue("Arcadia-Tag").isNullOrEmpty()) {
                    tagValue = attributes.getValue("Arcadia-Tag")
                }
            }
        } catch (ignored: IOException) {
        }
        this.version = ver ?: ""
        this.branch = branchValue ?: ""
        this.tag = tagValue ?: ""
    }
}
